src/utils.ts

```diff
@@ -54,19 +57,59 @@ export type ProcessWithOrderingParams<T> = {
  *
  * This same scenario is true for SQS FIFO queues, which will order messages
  * by MessageGroupId.
+ *
+ * When rejectOnError is false, the method will return the list
+ * of unprocessed items by listId. Callers will then have access to the error
+ * that caused the list to stop processing events, plus all the events
+ * that were not processed in the batch.
  */
 export const processWithOrdering = async <T>(
   params: ProcessWithOrderingParams<T>,
   process: (item: T) => Promise<void>,
 ) => {
-  const lists = Object.values(groupBy(params.items, params.orderBy));
+  const rejectOnError = params.rejectOnError ?? true;
+  const groupedItems = groupBy(params.items, params.orderBy);
+  const listIds = Object.keys(groupedItems);
+  const lists = Object.values(groupedItems);
+  const unprocessedRecordsByListId = zipObject<{ error: any; items: T[] }>(
+    listIds,
+    lists.map(() => ({ error: null, items: [] })),
+  );
```

Comment:
It's not clear to me why we need to keep this data in this "by list id" format, since we flatten it afterwards anyway.

Can we simplify this to just track a simple list of item+failure? e.g. to have this type: `{ item: T; error: any}[]`

Or, am I missing something.

Reply:
We have to remember that a failure in this scenario is not for a single item, but rather for a group of items that belong to the same list ID, hence why we keep this data structure.

This is one of the points from the SQS best practices for partial responses:
- `If you're using this feature with a [First-In-First-Out (FIFO) queue](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/FIFO-queues.html), your function should stop processing messages after the first failure and return all failed and unprocessed messages in batchItemFailures. This helps preserve the ordering of messages in your queue.`

It might seem like as the code is now, we're forcing non-FIFO users of processing things in order, but I think in reality that's not true, since `delta` was already generating a unique message group ID for each record that did not have `MessageGroupId` in its attributes, effectively splitting records into groups that only have a single item.

Does all that track? Happy to zoom about this if it's easier.

Reply:
That definitely tracks, and of course you're right. I was mixed up about the reading of the types here.